pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -181,183 +186,183 @@ public void stop() {
+   * <p>NOTE: the passed in IdealState may contain HLC segments if both HLC and LLC are configured.
    */
-  public void setupNewTable(TableConfig tableConfig, IdealState emptyIdealState)
-      throws InvalidConfigException {
-    final StreamConfig streamConfig = new StreamConfig(
-        tableConfig.getTableName(),
-        tableConfig.getIndexingConfig().getStreamConfigs());
-    int partitionCount = getPartitionCount(streamConfig);
-    List<String> currentSegments = getExistingSegments(tableConfig.getTableName());
-    // Make sure that there are no low-level segments existing.
-    if (currentSegments != null) {
-      for (String segment : currentSegments) {
-        if (!SegmentName.isHighLevelConsumerSegmentName(segment)) {
-          // For now, we don't support re-creating the low-level realtime segments
-          throw new RuntimeException("Low-level segments already exist for table " + tableConfig.getTableType());
-        }
-      }
+  public void setUpNewTable(TableConfig tableConfig, IdealState idealState) {
+    Preconditions.checkState(!_isStopping, "Segment manager is stopping");
+
+    String realtimeTableName = tableConfig.getTableName();
+    LOGGER.info("Setting up new LLC table: {}", realtimeTableName);
+
+    // Make sure all the existing segments are HLC segments
+    List<String> currentSegments = getAllSegments(realtimeTableName);
+    for (String segmentName : currentSegments) {
+      Preconditions.checkState(SegmentName.isHighLevelConsumerSegmentName(segmentName),
+          "Cannot set up new LLC table: %s with existing non-HLC segment: %s", realtimeTableName, segmentName);
     }
-    _flushThresholdUpdateManager.clearFlushThresholdUpdater(tableConfig);
-    if (!isConnected()) {
-      throw new RuntimeException(
-          "Lost zk connection while setting up new table " + tableConfig.getTableName() + " isConnected="
-              + isConnected());
+
+    _flushThresholdUpdateManager.clearFlushThresholdUpdater(realtimeTableName);
+
+    PartitionLevelStreamConfig partitionLevelStreamConfig = new PartitionLevelStreamConfig(tableConfig);
+    InstancePartitions instancePartitions = getConsumingInstancePartitions(tableConfig);
+    int numPartitions = getNumPartitions(partitionLevelStreamConfig);
+    int numReplicas = getNumReplicas(tableConfig, instancePartitions);
+
+    SegmentAssignment segmentAssignment = SegmentAssignmentFactory.getSegmentAssignment(_helixManager, tableConfig);
+    Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap =
+        Collections.singletonMap(InstancePartitionsType.CONSUMING, instancePartitions);
+
+    long currentTimeMs = getCurrentTimeMs();
+    Map<String, Map<String, String>> instanceStatesMap = idealState.getRecord().getMapFields();
+    for (int partitionId = 0; partitionId < numPartitions; partitionId++) {
+      String segmentName =
+          setupNewPartition(tableConfig, partitionLevelStreamConfig, partitionId, currentTimeMs, instancePartitions,
```

Comment:
I considered that, but that will make the method signature quite redundant. Since this is not performance sensitive path, I just cached the expensive operations such as extracting stream config from table config